File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultTests.java
package org.schabi.newpipe.extractor.services;

import org.schabi.newpipe.extractor.Image;
import org.schabi.newpipe.extractor.InfoItem;
import org.schabi.newpipe.extractor.ListExtractor;
import org.schabi.newpipe.extractor.Page;
import org.schabi.newpipe.extractor.StreamingService;
import org.schabi.newpipe.extractor.channel.ChannelInfoItem;
import org.schabi.newpipe.extractor.exceptions.ParsingException;
import org.schabi.newpipe.extractor.localization.DateWrapper;
import org.schabi.newpipe.extractor.playlist.PlaylistInfoItem;
import org.schabi.newpipe.extractor.stream.StreamInfoItem;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertEmptyErrors;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertGreaterOrEqual;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertIsSecureUrl;
import static org.schabi.newpipe.extractor.ExtractorAsserts.assertNotEmpty;
import static org.schabi.newpipe.extractor.StreamingService.LinkType;
import static org.schabi.newpipe.extractor.utils.Utils.isNullOrEmpty;

public final class DefaultTests {
    public static void defaultTestListOfItems(StreamingService expectedService, List<? extends InfoItem> itemsList, List<Throwable> errors) throws ParsingException {
        assertFalse(itemsList.isEmpty(), "List of items is empty");
        assertFalse(itemsList.contains(null), "List of items contains a null element");
        assertEmptyErrors("Errors during extraction", errors);

        for (InfoItem item : itemsList) {
            assertIsSecureUrl(item.url);

            final List<Image> thumbnails = item.thumbnails;
            if (!isNullOrEmpty(thumbnails)) {
                defaultTestImageCollection(thumbnails);
            }
            assertNotNull(item.infoType, "InfoItem type not set: " + item);
            assertEquals(expectedService.serviceId, item.serviceId, "Unexpected item service id");
            assertNotEmpty("Item name not set: " + item, item.name);

            if (item instanceof StreamInfoItem) {
                StreamInfoItem streamInfoItem = (StreamInfoItem) item;
                final String uploaderUrl = streamInfoItem.uploaderUrl;
                if (!isNullOrEmpty(uploaderUrl)) {
                    assertIsSecureUrl(uploaderUrl);
                    assertExpectedLinkType(expectedService, uploaderUrl, LinkType.CHANNEL);
                }

                final List<Image> uploaderAvatars = streamInfoItem.uploaderAvatars;
                if (!isNullOrEmpty(uploaderAvatars)) {
                    defaultTestImageCollection(uploaderAvatars);
                }

                assertExpectedLinkType(expectedService, streamInfoItem.url, LinkType.STREAM);

                if (!isNullOrEmpty(streamInfoItem.textualUploadDate)) {
                    final DateWrapper uploadDate = streamInfoItem.uploadDate;
                    assertNotNull(uploadDate,"No parsed upload date");
                }

            } else if (item instanceof ChannelInfoItem) {
                final ChannelInfoItem channelInfoItem = (ChannelInfoItem) item;
                assertExpectedLinkType(expectedService, channelInfoItem.url, LinkType.CHANNEL);

            } else if (item instanceof PlaylistInfoItem) {
                final PlaylistInfoItem playlistInfoItem = (PlaylistInfoItem) item;
                assertExpectedLinkType(expectedService, playlistInfoItem.url, LinkType.PLAYLIST);
            }
        }
    }

    private static void assertExpectedLinkType(StreamingService expectedService, String url, LinkType expectedLinkType) throws ParsingException {
        final LinkType linkTypeByUrl = expectedService.getLinkTypeByUrl(url);

        assertNotEquals(LinkType.NONE, linkTypeByUrl,
                "Url is not recognized by its own service: \"" + url + "\"");
        assertEquals(expectedLinkType, linkTypeByUrl,
                "Service returned wrong link type for: \"" + url + "\"");
    }

    public static void assertOnlyContainsType(ListExtractor.InfoItemsPage<? extends InfoItem> items, InfoItem.InfoType expectedType) {
        for (InfoItem item : items.getItems()) {
            assertEquals(expectedType, item.infoType,
                    "Item list contains unexpected info types");
        }
    }

    public static <T extends InfoItem> void assertNoMoreItems(ListExtractor<T> extractor) throws Exception {
        final ListExtractor.InfoItemsPage<T> initialPage = extractor.initialPage;
        assertFalse(initialPage.hasNextPage(), "More items available when it shouldn't");
    }

    public static void assertNoDuplicatedItems(StreamingService expectedService,
                                               ListExtractor.InfoItemsPage<InfoItem> page1,
                                               ListExtractor.InfoItemsPage<InfoItem> page2) throws Exception {
        defaultTestListOfItems(expectedService, page1.getItems(), page1.errors);
        defaultTestListOfItems(expectedService, page2.getItems(), page2.errors);

        final Set<String> urlsSet = new HashSet<>();
        for (InfoItem item : page1.getItems()) {
            urlsSet.add(item.url);
        }

        for (InfoItem item : page2.getItems()) {
            final boolean wasAdded = urlsSet.add(item.url);
            if (!wasAdded) {
                fail("Same item was on the first and second page item list");
            }
        }
    }

    public static <T extends InfoItem> ListExtractor.InfoItemsPage<T> defaultTestRelatedItems(ListExtractor<T> extractor) throws Exception {
        final ListExtractor.InfoItemsPage<T> page = extractor.initialPage;
        final List<T> itemsList = page.getItems();
        List<Throwable> errors = page.errors;

        defaultTestListOfItems(extractor.service, itemsList, errors);
        return page;
    }

    public static <T extends InfoItem> ListExtractor.InfoItemsPage<T> defaultTestMoreItems(ListExtractor<T> extractor) throws Exception {
        final ListExtractor.InfoItemsPage<T> initialPage = extractor.initialPage;
        assertTrue(initialPage.hasNextPage(), "Doesn't have more items");
        ListExtractor.InfoItemsPage<T> nextPage = extractor.getPage(initialPage.nextPage);
        final List<T> items = nextPage.getItems();
        assertFalse(items.isEmpty(), "Next page is empty");
        assertEmptyErrors("Next page have errors", nextPage.errors);

        defaultTestListOfItems(extractor.service, nextPage.getItems(), nextPage.errors);
        return nextPage;
    }

    public static void defaultTestGetPageInNewExtractor(ListExtractor<? extends InfoItem> extractor, ListExtractor<? extends InfoItem> newExtractor) throws Exception {
        final Page nextPage = extractor.initialPage.nextPage;

        final ListExtractor.InfoItemsPage<? extends InfoItem> page = newExtractor.getPage(nextPage);
        defaultTestListOfItems(extractor.service, page.getItems(), page.errors);
    }

    public static void defaultTestImageCollection(
            @Nullable final Collection<Image> imageCollection) {
        assertNotNull(imageCollection);
        imageCollection.forEach(image -> {
            assertIsSecureUrl(image.url);
            assertGreaterOrEqual(Image.HEIGHT_UNKNOWN, image.height,
                    "Unexpected image height: " + image.height);
            assertGreaterOrEqual(Image.WIDTH_UNKNOWN, image.width,
                    "Unexpected image width: " + image.width);
        });
    }
}

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultTests.kt
package org.schabi.newpipe.extractor.services

import org.junit.jupiter.api.Assertions
import org.schabi.newpipe.extractor.ExtractorAsserts
import org.schabi.newpipe.extractor.Image
import org.schabi.newpipe.extractor.InfoItem
import org.schabi.newpipe.extractor.InfoItem.InfoType
import org.schabi.newpipe.extractor.ListExtractor
import org.schabi.newpipe.extractor.ListExtractor.InfoItemsPage
import org.schabi.newpipe.extractor.StreamingService
import org.schabi.newpipe.extractor.StreamingService.LinkType
import org.schabi.newpipe.extractor.channel.ChannelInfoItem
import org.schabi.newpipe.extractor.exceptions.ParsingException
import org.schabi.newpipe.extractor.playlist.PlaylistInfoItem
import org.schabi.newpipe.extractor.stream.StreamInfoItem
import org.schabi.newpipe.extractor.utils.Utils.isNullOrEmpty
import java.util.function.Consumer

object DefaultTests {
    @Throws(ParsingException::class)
    fun defaultTestListOfItems(expectedService: StreamingService, itemsList: List<InfoItem?>?, errors: List<Throwable?>?) {
        Assertions.assertFalse(itemsList!!.isEmpty(), "List of items is empty")
        Assertions.assertFalse(itemsList.contains(null), "List of items contains a null element")
        ExtractorAsserts.assertEmptyErrors("Errors during extraction", errors)
        for (item in itemsList) {
            ExtractorAsserts.assertIsSecureUrl(item!!.url)
            val thumbnails: List<Image?> = item.thumbnails
            if (!isNullOrEmpty(thumbnails)) {
                defaultTestImageCollection(thumbnails)
            }
            Assertions.assertNotNull(item.infoType, "InfoItem type not set: $item")
            Assertions.assertEquals(expectedService.serviceId, item.serviceId, "Unexpected item service id")
            ExtractorAsserts.assertNotEmpty("Item name not set: $item", item.name)
            if (item is StreamInfoItem) {
                val streamInfoItem = item
                val uploaderUrl = streamInfoItem.uploaderUrl
                if (!isNullOrEmpty(uploaderUrl)) {
                    ExtractorAsserts.assertIsSecureUrl(uploaderUrl)
                    assertExpectedLinkType(expectedService, uploaderUrl, LinkType.CHANNEL)
                }
                val uploaderAvatars = streamInfoItem.uploaderAvatars
                if (!isNullOrEmpty(uploaderAvatars)) {
                    defaultTestImageCollection(uploaderAvatars)
                }
                assertExpectedLinkType(expectedService, streamInfoItem.url, LinkType.STREAM)
                if (!isNullOrEmpty(streamInfoItem.textualUploadDate)) {
                    val uploadDate = streamInfoItem.uploadDate
                    Assertions.assertNotNull(uploadDate, "No parsed upload date")
                }
            } else if (item is ChannelInfoItem) {
                assertExpectedLinkType(expectedService, item.url, LinkType.CHANNEL)
            } else if (item is PlaylistInfoItem) {
                assertExpectedLinkType(expectedService, item.url, LinkType.PLAYLIST)
            }
        }
    }

    @Throws(ParsingException::class)
    private fun assertExpectedLinkType(expectedService: StreamingService, url: String?, expectedLinkType: LinkType) {
        val linkTypeByUrl = expectedService.getLinkTypeByUrl(url)
        Assertions.assertNotEquals(LinkType.NONE, linkTypeByUrl,
                "Url is not recognized by its own service: \"$url\"")
        Assertions.assertEquals(expectedLinkType, linkTypeByUrl,
                "Service returned wrong link type for: \"$url\"")
    }

    fun assertOnlyContainsType(items: InfoItemsPage<out InfoItem?>?, expectedType: InfoType?) {
        for (item in items!!.items!!) {
            Assertions.assertEquals(expectedType, item!!.infoType,
                    "Item list contains unexpected info types")
        }
    }

    @Throws(Exception::class)
    fun <T : InfoItem?> assertNoMoreItems(extractor: ListExtractor<T>?) {
        val initialPage = extractor!!.initialPage
        Assertions.assertFalse(initialPage!!.hasNextPage(), "More items available when it shouldn't")
    }

    @Throws(Exception::class)
    fun assertNoDuplicatedItems(expectedService: StreamingService,
                                page1: InfoItemsPage<InfoItem?>?,
                                page2: InfoItemsPage<InfoItem?>?) {
        defaultTestListOfItems(expectedService, page1!!.items, page1.errors)
        defaultTestListOfItems(expectedService, page2!!.items, page2.errors)
        val urlsSet: MutableSet<String?> = HashSet()
        for (item in page1.items!!) {
            urlsSet.add(item!!.url)
        }
        for (item in page2.items!!) {
            val wasAdded = urlsSet.add(item!!.url)
            if (!wasAdded) {
                Assertions.fail<Any>("Same item was on the first and second page item list")
            }
        }
    }

    @Throws(Exception::class)
    fun <T : InfoItem?> defaultTestRelatedItems(extractor: ListExtractor<T>?): InfoItemsPage<T?>? {
        val page = extractor!!.initialPage
        val itemsList = page!!.items
        val errors = page.errors
        defaultTestListOfItems(extractor.service, itemsList, errors)
        return page
    }

    @Throws(Exception::class)
    fun <T : InfoItem?> defaultTestMoreItems(extractor: ListExtractor<T>?): InfoItemsPage<T?>? {
        val initialPage = extractor!!.initialPage
        Assertions.assertTrue(initialPage!!.hasNextPage(), "Doesn't have more items")
        val nextPage = extractor.getPage(initialPage.nextPage)
        val items = nextPage!!.items
        Assertions.assertFalse(items!!.isEmpty(), "Next page is empty")
        ExtractorAsserts.assertEmptyErrors("Next page have errors", nextPage.errors)
        defaultTestListOfItems(extractor.service, nextPage.items, nextPage.errors)
        return nextPage
    }

    @Throws(Exception::class)
    fun defaultTestGetPageInNewExtractor(extractor: ListExtractor<out InfoItem?>?, newExtractor: ListExtractor<out InfoItem?>) {
        val nextPage = extractor!!.initialPage!!.nextPage
        val page = newExtractor.getPage(nextPage)
        defaultTestListOfItems(extractor.service, page!!.items, page.errors)
    }

    fun defaultTestImageCollection(
            imageCollection: Collection<Image?>?) {
        Assertions.assertNotNull(imageCollection)
        imageCollection!!.forEach(Consumer { image: Image? ->
            ExtractorAsserts.assertIsSecureUrl(image!!.url)
            ExtractorAsserts.assertGreaterOrEqual(Image.HEIGHT_UNKNOWN.toLong(), image.height.toLong(),
                    "Unexpected image height: " + image.height)
            ExtractorAsserts.assertGreaterOrEqual(Image.WIDTH_UNKNOWN.toLong(), image.width.toLong(),
                    "Unexpected image width: " + image.width)
        })
    }
}
